velox/common/base/Status.h

```diff
@@ -0,0 +1,459 @@
+/// pointer. For failure cases, it allocates an external object containing the
+/// StatusCode and error message. This keeps the object compact and prevents
+/// allocations in the common success case. The same strategy is used in Status
+/// object from other well-known libraries like Apache Arrow, RocksDB, Kudu, and
+/// others.
+///
+/// Simple usage:
+///
+///  Status operation() {
+///    if (noMoreMemory) {
+///      return Status::OutOfMemory("Not enough memory to run 'operation'!");
+///    }
+///    return Status::OK();
+///  }
+///
+/// Call site:
+///
+///  auto status = operation();
+///  if (status.ok()) {
+///    ...
+///  } else if (status.isOutOfMemory()) {
+///    (gracefully handle out of memory)
+///  } else if (status.isNotImplemented()) {
+///    (and so on)
+///  }
+///
+/// Other common usage patterns:
+///
+/// The same logic above can be implemented using helper macros:
+///
+///  Status operation() {
+///    VELOX_RETURN_IF(noMoreMemory, Status::OutOfMemory(
+///       "Not enough memory to run 'operation'!"));
+///    ...
+///    return Status::OK();
+///  }
+///
+/// To ensure operations succeed (or if not, return the same status from the
+/// current function):
+///
+///  ...
+///  VELOX_RETURN_NOT_OK(operation1());
+///  VELOX_RETURN_NOT_OK(operation2());
+///  VELOX_RETURN_NOT_OK(operation3());
+///  ...
+
+enum class StatusCode : int8_t {
+  kOK = 0,
+  kUserError = 1,
+  kTypeError = 2,
+  kIndexError = 3,
+  kKeyError = 4,
+  kAlreadyExists = 5,
+
+  kOutOfMemory = 6,
+  kIOError = 7,
+
+  kCancelled = 8,
+  kInvalid = 9,
+  kUnknownError = 10,
+
+  kNotImplemented = 11,
+};
+std::string_view toString(StatusCode code);
+
+class [[nodiscard]] Status {
+ public:
+  // Create a success status.
+  constexpr Status() noexcept : state_(nullptr) {}
+
+  ~Status() noexcept {
+    if (FOLLY_UNLIKELY(state_ != nullptr)) {
+      deleteState();
+    }
+  }
+
+  Status(StatusCode code, std::string msg);
+
+  // Copy the specified status.
+  inline Status(const Status& s);
+  inline Status& operator=(const Status& s);
+
+  // Move the specified status.
+  inline Status(Status&& s) noexcept;
+  inline Status& operator=(Status&& s) noexcept;
+
+  inline bool operator==(const Status& other) const noexcept;
+  inline bool operator!=(const Status& other) const noexcept {
+    return !(*this == other);
+  }
+
+  // AND the statuses.
+  inline Status operator&(const Status& s) const noexcept;
+  inline Status operator&(Status&& s) const noexcept;
+  inline Status& operator&=(const Status& s) noexcept;
+  inline Status& operator&=(Status&& s) noexcept;
+
+  inline friend std::ostream& operator<<(std::ostream& ss, const Status& s) {
+    return ss << s.toString();
+  }
+
+  /// Return a success status.
+  static Status OK() {
+    return Status();
+  }
+
+  /// Return an error status for out-of-memory conditions.
+  template <typename... Args>
+  static Status UserError(Args&&... args) {
```

Comment:
these names do not follow coding guidelines that says method names should be lower camel case

Reply:
I've debated about these as well. These are meant to be used to as factories, like:

```
auto status1 = Status::UserError("error msg");
auto status2 = Status::OK();
auto status3 = Status::TypeError("error msg");
```

maybe I'm just too used to arrow, but it sorts of looks weird to have them as lower camel case:

```
auto status1 = Status::userError("error msg");
auto status2 = Status::ok();
auto status3 = Status::typeError("error msg");
```

I don't have a strong opinion though. Happy to make the change. 

@mbasmanova  what do you think?

Reply:
I don't mind either way. If you decide to keep it as is, maybe, add a comment to explain that deviation from coding standard is intentional.